Peritiae

pplans opened the pull request.

Lysoun commented on template.json:
```diff
@@ -10,41 +10,77 @@
+                    "value": 14,
+                    "defaultRoll": "d10+d4"
+                }
+            },
```
Pretty sure that peritiae have minimum values (maybe 2? I don't really remember). I don't see the use of defining a default roll at that point because it will never be used, since users do pick their rolls every time they roll the peritia. Do you have an example of the use of such an attribute?

pplans replied:
Good catch for the minimum, but as this is a general rule, maybe we can put it in the system directly, like upstairs in the object. For the default roll, I think its a good idea since I observed that players use to roll the same dices often, so it will allow to use a default roll with a button and it will help speed up things. And when players will want to change it they could aside.

Lysoun commented on template.json:
```diff
@@ -10,41 +10,77 @@
-                "background",
-                "resources"
+                "common"
             ],
```
I do think more of these categories should have peritiae. In combat, the player rolls De Bello against the enemy De Corpore (or De Bello? not sure) thus the enemy should have such an attribute.

Lysoun commented on templates/items/shield.html:
```diff
@@ -22,7 +22,7 @@ <h4 class="item-type">{{itemType}}</h4>
         <div class="tab flexrow active" data-group="primary" data-tab="description">
             <p>Description</p>
-            {{editor content=data.description.value target="data.description.value" button=true owner=owner editable=editable}}
+            {{editor content=data.description target="data.description" button=true owner=owner editable=editable}}
```
I think there is quite some duplication here. Isn't there a way to create a part "description" that you would reuse every time that you want to put a description? It could go in a file in templates/parts.

pplans replied:
Definitely, we can make a part for the description to share it among the entities

Lysoun commented on templates/actors/limited-sheet.html:
```diff
@@ -14,8 +14,7 @@ <h1 class="charname">
         <div class="tab biography">
-            <p>Biography</p>
-            {{editor content=data.details.biography.value target="data.details.biography.value" button=true owner=owner editable=editable}}
+            {{editor content=data.biography target="data.biography" button=true owner=owner editable=editable}}
```
There is duplication of the biography. I would extract that to a part and reuse it.

pplans replied:
As per the description

pplans replied:
I went through it again, since the biography is only taking one line to set, I prefer avoiding splitting everything in subtemplates as I don't know the cost to invoke a subtemplate yet. I prefer use subtemplates when things get a bit complicated, for instance maybe the Peritia ? I don't know to be honest

Lysoun commented on templates/actors/custos-sheet.html:
```diff
@@ -45,6 +45,20 @@ <h4 class="attribute-name box-title">{{ localize "LexArcana.Pietas" }}</h4>
+                    <li class="peritia">
+                        <h4 for="data.peritiae.{{peritiaKey}}.defaultValue" class="peritia-name rollable">{{peritiaValue.label}}</h4>
+                        <input name="data.peritiae.{{peritiaKey}}.value" value="{{peritiaValue.value}}" data-dtype="number"/>
+                        Default Roll: <input name="data.peritiae.{{peritiaKey}}.defaultRoll" value="{{peritiaValue.defaultRoll}}" data-dtype="number"/>
```
As stated before, I don't see the use of such a default roll.

Lysoun commented on lang/en.json:
```diff
@@ -1,22 +1,28 @@
+  "LexArcana.PeritiaDeSocietate": "De Societate",
+  "LexArcana.PeritiaDeMagia":     "De Magia",
+  "LexArcana.PeritiaDeScientia":  "De Scientia",
+  "LexArcana.PeritiaDeCorpore":   "De Corpore",
```
De Corpore is a value for a peritia, thus it should be named like that: ```LexArcana.Peritia.DeCorpore```. The same goes for all peritiae

pplans replied:
Why not, I was following examples around, I'll take a look :)

Lysoun commented on module/actor/sheets/custos.js:
```diff
@@ -9,17 +9,75 @@ import {LexArcana} from '../../config.js';
+      editable: this.isEditable,
+      cssClass: isOwner ? "editable" : "locked",
+      isCharacter: this.entity.data.type === "custos",
+      isNPC: this.entity.data.type !== "custos",
```
A custos may be a NPC given by the GM to the players to help them during the adventure. If those booleans serve the purpose of distinguishing the player characters and the NPCs controlled by the GM, I don't think this condition is going to work.

pplans replied:
Yeah that can't last, for the moment we have Custos and Friendly to distinguish them but clearly we can set a boolean there with a checkbox

Lysoun commented on module/utils.js:
```diff
@@ -0,0 +1,9 @@
+LexArcanaUtils.ObjectToArray = function _ObjectToArray(obj)
+{
+	return Object.values(obj);
+}
```
How about this simpler way to write it: 
```javascript
export const LexArcanaUtils = {
    /* -------------------------------------------- */
    /*                  Tools                       */
    /* -------------------------------------------- */
    function ObjectToArray(obj)
    {
	return Object.values(obj);
    }
}
```

pplans replied:
same, and its missing static keyword

pplans replied:
Hum after reflexion, it would be better with the class and static keywords, same goes for dice.js

Lysoun commented on module/dice.js:
```diff
@@ -0,0 +1,71 @@
+	for(let current = 0;current<numDice;++current)
+	{
+		let filterFaces = currentMaxFaces/(numDice-current);
+		let nextFace = validDices.find(value => value<=filterFaces);
```
const

pplans replied:
fixed

Lysoun commented on module/dice.js:
```diff
@@ -0,0 +1,71 @@
+	let expression = "";
+	for(let current = 0;current<numDice;++current)
+	{
+		let filterFaces = currentMaxFaces/(numDice-current);
```
const

pplans replied:
fixed

Lysoun commented on module/dice.js:
```diff
@@ -0,0 +1,71 @@
+	}
+	let i = 0;
+	let nMaxIterations = 12;
+	while(sumDiceFaces<maxFaces && nMaxIterations-->0)
```
```-->``` ? Never saw this, can you explain it to me?

pplans replied:
easy, -- first, then >, but you're right, better prefix decrement

Lysoun commented on module/dice.js:
```diff
@@ -0,0 +1,71 @@
+	let nMaxIterations = 12;
+	while(sumDiceFaces<maxFaces && nMaxIterations-->0)
+	{
+		let filterFaces = (maxFaces-sumDiceFaces)+diceFaces[i];
```
const

pplans replied:
fixed next changelist

Lysoun commented on styles/LexArcana.css:
```diff
@@ -0,0 +1,236 @@
+{
+    font-family: Herculanum;
+    font-size: 14px;
+    font-weight: 700;
```
Isn't a unit missing?

pplans replied:
fixed next changelist

Lysoun commented on styles/LexArcana.css:
```diff
@@ -0,0 +1,236 @@
+, .tox
+{
+    padding: 5px;
+	background-color: #e0cfaa;
```
indentation

pplans replied:
not in visual studio code nor notepad++ 🤔

Lysoun commented on styles/LexArcana.css:
```diff
@@ -0,0 +1,236 @@
+{
+    list-style: none;
+    margin-top: 15px;
+    margin-bottom: auto;
```
I know why you used such a ```margin-bottom``` but I believe you could have the same result using ```align-items: flex-start``` since you are inside a flex block. See the [documentation](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Aligning_Items_in_a_Flex_Container) for more information

pplans replied:
I tried, I had to add some lines more and it complexified a lot the whole css making it even uglier and difficult to read. So I'll stick with margin-bottom, simple and does the job as intended

Lysoun commented on template.json:
```diff
@@ -1,69 +1,121 @@
+                    "deBello":
+                    {
+                        "value": 14,
+                        "specialties": {},
```
Aren't specialties arrays? Didn't you have a ton of issues because of specialties becoming objects and couldn't that be the problem?

pplans replied:
It f**** up everything if I do and I have to add a lot of more css lines. So I maintain the margin-bottom that works fine

Lysoun replied:
Mhm I think you didn't answer to the right comment here

Lysoun reviewed (changes requested):
The file module/actor/entity.js only contains the declaration of the class ```LexArcanaActor```. Why isn't it its name? 

Some general comments: 
* you are not consistent on the use of single quotes or double quotes for your string declarations. The norm would be using single quotes.
* using binary operators, you sometimes put a space before and after and sometimes you don't. I find it way easier to read when there are always a space before and another one after the binary operator

pplans commented on lang/en.json:
```diff
@@ -1,22 +1,42 @@
+  "LexArcana.Range":              "Range",
+  "LexArcana.Difficulty":         "Difficulty",
+  "LexArcana.Feat":               "Feat",
+  "LexArcana.Parry":              "Parry"
```
Because the Peritiae work quite differently, they are two steps translation. Why exactly I don't remember well, its probably because we started the project with the D&D 5e system files and we took them as an example

pplans commented on module/actor/entity.js:
```diff
@@ -69,4 +71,79 @@ export default class LexArcanaActor extends Actor
+    {
+        const peritia = this.data.data.peritiae[peritiaId];
+        let idx = LexArcanaUtils.ObjectToArray(peritia.specialties).findIndex(item => item.name===speName);
+        return peritia.specialties[idx];
```
Fixed in the next changelist

pplans commented on module/actor/entity.js:
```diff
@@ -69,4 +71,79 @@ export default class LexArcanaActor extends Actor
+    async removePeritiaSpecialty(peritiaId, key)
+    {
+        let currentSpecialties = duplicate(this.data.data.peritiae?.[peritiaId]?.specialties ?? []);
+        currentSpecialties = LexArcanaUtils.ObjectToArray(currentSpecialties).filter(item => item.name!==key);
```
Because here we need the complete list, so we will add code somewhere else to save code here. For the moment its useless since we only need this here, but maybe later if we need to do the same thing we can come up with something

pplans commented on module/actor/entity.js:
```diff
@@ -69,4 +71,79 @@ export default class LexArcanaActor extends Actor
+        {
+            currentSpecialties[idx].defaultRoll = newExpression;
+            await super.update({[`data.peritiae.${peritiaId}.specialties`]: [...currentSpecialties] });
+        }
```
We can deport the retrieving of the specialty list but anything else is logic specific here

pplans commented on module/actor/entity.js:
```diff
@@ -69,4 +71,79 @@ export default class LexArcanaActor extends Actor
+        if(currentSpecialties[idx]!==undefined)
+        {
+            currentSpecialties[idx].defaultRoll = newExpression;
+            await super.update({[`data.peritiae.${peritiaId}.specialties`]: [...currentSpecialties] });
```
It's wrapping and adding other asynchronous calls to others. Its overweight for performance to gain too few visibility

pplans commented:
I renamed both entity.js files actor and item one as respectively actor.js and item.js, it was a inherited from the D&D5e system but I always thought it was bad naming. Thing is, it should not have been done in this branch probably.

Also replaced every possible double quotes into single quotes

pplans commented on module/actor/entity.js:
```diff
@@ -69,4 +71,79 @@ export default class LexArcanaActor extends Actor
+    /*	Roll Dices
+     */
+
+    /* -------------------------------------------- */
```
nah just those comments that I don't maintain that much, I miss #pragma region from VS C++ :(

pplans commented on module/actor/entity.js:
```diff
@@ -69,4 +71,79 @@ export default class LexArcanaActor extends Actor
+    roll(expression, info="")
+    {
+        let rollMode = game.settings.get('core', 'rollMode');
+        let message = {speaker: {actor: this.id }, content: info+" "+expression, blind: rollMode === 'blindroll' };
```
iup!

pplans commented on module/actor/entity.js:
```diff
@@ -69,4 +71,79 @@ export default class LexArcanaActor extends Actor
+    {
+        let rollMode = game.settings.get('core', 'rollMode');
+        let message = {speaker: {actor: this.id }, content: info+" "+expression, blind: rollMode === 'blindroll' };
+        if(expression.match("^(([1-9][0-9]*)?d([34568]|(?:10)|(?:12)|(?:20)))(\\+(([1-9][0-9]*)?d([34568]|(?:10)|(?:12)|(?:20))))*"))
```
Why not!

pplans commented on module/dice.js:
```diff
@@ -0,0 +1,71 @@
+		if(nextFace<(currentMaxFaces+1))
+		{
+		currentMaxFaces -= nextFace;
+		expression += (expression!==""?"+":"")+"1d"+nextFace;
```
Already done :)

pplans commented on module/dice.js:
```diff
@@ -0,0 +1,71 @@
+
+LexArcanaDice.ComputeExpressionBalanced = function(numDice, maxFaces)
+{
+	const validDices = [20, 12, 10, 8, 6, 5, 4, 3];
```
For no available reason, what is the difference ? Like pros and cons ? Also, they should be static

pplans replied:
It does not seem to add anything if I believe results on google so no real reasons to change it
